### DISCUSSION TITLE
What is the current state of mobile development with Avalonia ?

### DISCUSSION THREAD

**Discussion by buzzware** [AvaloniaUI/Avalonia#13257] [Q&A]
I'm trying to make a long term decision for mobile development with C#, coming from Xamarin Forms.
I see that Avalonia has only supported mobile since June, and many mobile specific elements don't seem to be there, or there are many third party options and its not clear which will be around for the long term.
In particular I'm thinking of stack navigation, hamburger with a fly out menu, title bar with back arrow, title & action buttons etc.
Is there a happy path for mobile development yet?

**Comment by buzzware** [comment]
I found 
[https://github.com/AvaloniaUI/Avalonia.Labs/pull/9](https://github.com/AvaloniaUI/Avalonia.Labs/pull/9) & [https://github.com/AvaloniaUI/Avalonia/issues/10557](https://github.com/AvaloniaUI/Avalonia/issues/10557) for mobile page navigation UI but it isn't complate and seems a little stagnant

**Reply by wieslawsoltes** [reply]
Take a look at https://github.com/AvaloniaInside/Shell

**Reply by buzzware** [reply]
Thanks @wieslawsoltes, that's very promising. Is that destined for the main product, or just experimental?
There are glitches, but it covers what I was talking about.
Its also a good demo of a scrolling list. Its not exactly buttery smooth, and there is a significant freak out when I scroll to the top on an Android device, but the performance is more than good enough even on this older industrial device.

**Reply by timunie** [reply]
for performance don't use debug builds. On low power devices AOT builds can help a lot. 

The shell view itself was not build by us afaik, so I don't think it will get merged to Avalonia itself. However you can check out Avalonia.Labs as pointed out earlier.

**Reply by buzzware** [reply]
I did 
Shell/src/Example/ShellExample/ShellExample.Android dotnet build --configuration Release
and ran it on Android emulator, Pixel 2 and an old Honeywell Android.
I also tried debug builds on Mac & Web.
Web was the only one where I thought the scrolling was unacceptable.

**Reply by BAndysc** [reply]
@buzzware it looks like this example doesn't use native AOT which makes a huge difference on web and probably on mobile too.

Take a look at: https://github.com/teobugslayer/AvaloniaCoreRTDemo/blob/master/src/AvaloniaCoreRTDemo.csproj (`PublishAot`).

**Comment by cok666n** [comment]
Labs and other packages aside, I feel like the question was not answered. 

Can we, as devs, expect Avalonia to become/stay a valid option for mobile app development in C# in the years to come?

Especially in the context of MAUI being a bit of a letdown (for us anyway) and Xamarin going away soon, this question will need to be answered as we will need to make choices for the long term. And I could not find an answer to that question myself either.

**Reply by thevortexcloud** [reply]
I can't speak for the Avalonia devs, but I suspect it will remain supported as long as Microsoft keeps supporting Xamarin (that's Xamarin itself, not Xamarin forms).

**Reply by MikeCodesDotNET** [reply]
We're continuing to work on improving our mobile support and are committed to mobile for the long term. The basic building blocks are mostly there, but they're rough around the edges, which is something we want to fix.

**Reply by timunie** [reply]
Avalonia listens to the users way more than other libs do. So as long as the community is interested, development goes further. And right now we see increased interest in mobile development. 

Anything missing for you? Reach out to us, let us know your thoughts and needs.

**Reply by cok666n** [reply]
That is indeed good news. Thanks for the info, and for considering users in your choices, it is truly appreciated.

**Reply by buzzware** [reply]
> We're continuing to work on improving our mobile support and are committed to mobile for the long term. The basic building blocks are mostly there, but they're rough around the edges, which is something we want to fix.

@MikeCodesDotNET Your answer above is only mildly encouraging to me. It seems to me that mobile is the main game. Perhaps I'm not connected enough to large enterprise (I'm working for one now, but my experience is mostly startups).
There are many apps that will always be proper HTML web apps - lets exclude them. Then there are proper desktop apps like MS Office - exclude them.
What we are left with is a lot of custom apps for which cross platform frameworks offer the ability to build once across browser, desktop & mobile. The ideal way to develop them is using a mobile-responsive framework eg. a hamburger menu that is always present or normally hidden depending on screen width, so essentially a *mobile* app.
I reckon Avalonia should be mobile-first and needs things like stack navigation, hamburger menu, routing, responsiveness (like Bootstrap) etc as soon as they can be done properly.

**Reply by stevemonaco** [reply]
> I reckon Avalonia should be mobile-first and needs things like stack navigation, hamburger menu, routing, responsiveness (like Bootstrap) etc as soon as they can be done properly.

I have to disagree here on the mobile-first part as someone who doesn't care about mobile at all. I write desktop apps and maybe someday web via Avalonia WASM. Features like stack navigation and routing can be trivial to implement for simple apps (20-50 LoC each?), but become significantly complex when you start trying to cover everybody's use cases. You can check out UWP's `NavigationView` for some idea. Hamburger menu shouldn't be hard to implement either. I'm not saying that these should never be supported, but these components should be straightforward to write app-specific versions for experienced XAML teams and aren't a make-or-break for a project. I might be more concerned about the lack of themes that mobile users are accustomed to like Material3.

Responsiveness to screen dimensions is a harder problem. The Avalonia.Xaml.Behaviors package includes an implementation and you can see a [sample here](https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors/blob/master/samples/BehaviorsTestApplication/Views/Pages/AdaptiveBehaviorView.axaml). This is something that should probably get more visibility, but I'm not sure how. The team has mentioned interest in implementing a version of media queries, so that would become the default approach if designed and implemented.

**Reply by jp2masa** [reply]
> The team has mentioned interest in implementing a version of media queries, so that would become the default approach if designed and implemented.

There is a PR for that: #7938.

**Reply by buzzware** [reply]
@stevemonaco Google Trends shows the relative interest in desktop vs mobile development. 
<img width="880" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/28285/cba25ef0-8971-4ad1-8418-cf19d813bdb5">

Many developers will be bringing their apps to Avalonia from Xamarin, which has a built in mobile hamburger menu, as do MAUI, Flutter and Uno. React Native has https://reactnavigation.org.
The hamburger menu is so foundational now that few would be happy with building one from scratch when choosing a framework. It also suggests a lack of commitment to mobile use cases.

**Reply by buzzware** [reply]
This is a pretty good effort, and provides most of what I need, but it would be much better to have similar built in and supported
https://github.com/AvaloniaInside/Shell
@OmidID

**Reply by OmidID** [reply]
@buzzware Thanks for sharing. In fact I agree with you something such as shell is important today. Most of the platform are currently have similar things builtin. This is the reason I develop the Shell.

**Comment by buzzware** [comment]
To put a fine point on it:
1) Many enterprise developers are committed to the Microsoft stack
2) There are many industrial applications for which a true mobile experience is required - not webview apps
3) MAUI seems to be failing - publicly visible development has significantly slowed this year and there are many serious articles and comments about quality issues
4) 1-3 leaves us with Uno and Avalonia. I can't imagine another contender appearing now and overtaking these two.
5) I've tried Uno with a lot of assistance from them and I've hit a lot of issues getting started. They may get fixed. They seem ahead of Avalonia on mobile and in their standard controls. They have a large mobile app agency nventiv behind them, but that agency is using and hiring for Flutter too (apparently not hiring for dotnet or Uno). Uno is more ambitious - using native controls and reimplementing WinUI cross-platform. Uno doesn't have a good beginner experience yet.
6) Avalonia has potential for fast growth because it is built on simply dotnet-<platform> and SkiaSharp, both solid, mature and maintained by microsoft without the challenge of native controls etc of MAUI. There is a lot more common code across platforms in both the UI framework and the app this way. You can debug down to the drawing routines in C#. And that drawing library is very capable - curves, gradients etc. It might not perform quite as fast as native, but Flutter (also using Skia until very recently) has shown that it is more than good enough for enterprise apps. Flutter is also good enough for a Fortune 500 resources company I worked for to build a suite of internal apps using Flutter for Web and not mobile.
7) Right now and over the next year there is probably a lot of Xamarin Forms apps and developers considering and reconsidering their choices, which will grow especially with the EOL May 2024. Many will be lost to Flutter & React Native, but many won't, so I expect we'll see rapid growth in Uno and/or Avalonia unless both quality & dev experience dramatically improves with MAUI soon. Flutter is my favourite, but one thing that keeps me interested in dotnet mobile is that for anything crossplatform and advanced eg. using background services, location, bluetooth, hardware etc you have a single language for all platforms and can call the platform APIs directly with compiled code. No other cross platform framework has that.
8) Both Uno and Avalonia need to get it together fast because the next 12 months will likely determine their future. In particular, online community is sorely lacking for both. This forum is something, but is it the long term plan? If not, migrate now because change gets harder and more painful as you grow. I suggest https://www.discourse.org (I hate Discord). The August Q&A was great, but what happened to the September one ? How about a podcast, a component website like https://fluttergems.dev etc etc. @MikeCodesDotNET

**Reply by MikeCodesDotNET** [reply]
Thank you for your detailed insights. While I won't address all points, I'd like to clarify a few. While many enterprises rely on Microsoft technologies, Avalonia's unique capabilities have proven to fit specific needs better, making the .NET ecosystem more of a cooperative space than a zero-sum game. 

Contrary to the idea that the next 12 months will be make-or-break for technologies like Avalonia, I believe the landscape is more nuanced. While mobile is a focus, our community's interest in Desktop, Embedded, and Web remains strong.

Regarding community engagement, our team values it highly. Missing last month's community call was an exception, and we're working on scheduling another one soon. GitHub remains the best platform for discussions despite considering options like Discourse. We also have a Telegram chat, which is very active if you want to join a vibrant community. 

Lastly, on the topic of resource directories akin to FlutterGems, it's in the pipeline but constrained by the size of our team and the long list on our roadmap. Given our team's size, we must balance the community needs with revenue-generating activities carefully.

**Reply by buzzware** [reply]
Thanks for replying. In the next 12 months enterprise (ie with funds) Xamarin Forms *mobile* apps will be evaluating MAUI, Uno & Avalonia first (because they can keep their non-visual logic in C#) but if they don't appear viable they will be going to Flutter or React Native. None of the C# options are super impressive right now if you expect even the level of polish they came to Xamarin Forms for.
Once they've gone to Flutter or React Native, they're not coming back.

**Reply by timahrentlov** [reply]
I agree.  

The window of opportunity to woo MAUI and Xamarin Forms mobile developers is still there. But for how much longer? We all know the "Dark Side" is calling for those cross-platform devs, and once they turn for mobile apps, so does their future desktop apps. I suppose a lot of energy is being used on development that benefits XPF, one way or another. For good reasons. But does that point to the future? Will WPF for desktop capture the heart and minds of new devs? Better than "you know who" ?

**Comment by workgroupengineering** [comment]
Hi @MikeCodesDotNET , 

The main problem is not having a public roadmap for planning your investment. This applies to companies but also to contributors like me, who donate their time on features that are not a high priority for the Team. Personally, being part of a nano company, contributing via PR is the only way to reciprocate the excellent work done.

**Reply by MikeCodesDotNET** [reply]
Thanks for your input. 

We have a roadmap, but it's currently only available to paying customers because our plans are fluid, and we're careful about setting expectations that might change. 

To mitigate any mismatch in priorities, we've established [contributor guidelines](https://github.com/AvaloniaUI/Avalonia/blob/master/CONTRIBUTING.md) that recommend direct engagement with our core team.

**Reply by workgroupengineering** [reply]
Thank you for Avalonia,

> To mitigate any mismatch in priorities, we've established [contributor guidelines](https://github.com/AvaloniaUI/Avalonia/blob/master/CONTRIBUTING.md) that recommend direct engagement with our core team.

and that's the problem. The Team is small and cannot answer all questions/proposals, because it is rightly busy with paying customers.
Having a public roadmap, the community's efforts could also be channeled through hacktoberfest-type initiatives.

**Reply by buzzware** [reply]
Why would you want to only upset your paying customers with fluid plans?
As a potential paying customer, seeing a roadmap would give me a bit extra confidence to become one.
Even if I had no intention of paying, a roadmap would give me a bit extra confidence to get involved.

**Reply by MikeCodesDotNET** [reply]
Largely because we've NDAs in place and long-term relationships with our customers. 

We will share our roadmap publicly when its ready.

**Reply by wsficke** [reply]
> Largely because we've NDAs in place and long-term relationships with our customers.

As a prospective customer, I appreciate the partnership attitude.  This is key.  However, @buzzware's point is not moot.  Avalonia's public facing documentation is lacking in many areas; product strategy and alignment is just one area.  To capture developer mind share, Avalonia must pass customer evaluation.

Currently, Avalonia is the leading .NET MVVM framework.  But Avalonia does not have sector dominance.  There's still a lot of impedance for new development.  However, the other options are either a lot more expensive, obsolete, or not quite there.

The future of Avalonia is rightly on the prospective customer's mind and integral to the value proposition.

Paying customers with NDAs should be prioritized by the core team.  This is the only viable long-term strategy for such a kind of project.  However, attracting paying customers is also necessary.

You have to give enough juice so we know it's good, without spilling the beans on your NDAs :)

**Reply by MikeCodesDotNET** [reply]
Hey @wsficke, 

I completely understand the value you place on understanding our direction, and I appreciate the feedback. You raise valid points about 'developer mindshare'! 

We're aware of areas where our documentation needs improvement and are actively working to improve its quality. Last year, we invested a five-figure sum in a technical writer, which helped a little. Funding this continued work is expensive, so we set our GitHub sponsorship goal to cover the cost of a part-time technical writer, which has reached 22% of the target. The more sponsorship the project receives, the easier it becomes to fund improvements. I [wrote last year about how we spend our earnings](https://avaloniaui.net/Blog/balancing-growth-and-sustainability-avalonia-s-financial-evolution), and I hope you can appreciate that we're not a big, VC-backed startup with cash to burn. We have to be incredibly deliberate when investing in Avalonia's future. 

Our current priority is ensuring Avalonia solves the core problems of our paying customers, as their support is essential for the framework's long-term sustainability. Their valuable input directly shapes the trajectory of our development. That isn't to say we disregard the community's needs. In fact, our active community engagement poses a unique commercial challenge. Companies often find they can get timely support through issues and discussions on GitHub, making our support agreements' value proposition more challenging to articulate. 

While we don't provide a public roadmap, rest assured that there's exciting work happening behind the scenes. We're focused on enhancing performance, expanding platform compatibility, and building incredibly exciting new things. We're committed to sharing news of these improvements as they become ready for release.

If you're seriously considering Avalonia for your next project and want to explore becoming a paying customer, please reach out directly. I'd happily schedule a call to discuss your specific needs and how a partnership could benefit you. While I can't share our detailed roadmap publicly, I'm always open to discussing our direction with potential customers.

**Comment by kevkov** [comment]
Hi, would it possible to provide a list of those controls that have or have not (whichever makes most sense) been developed with mobile/touch/gestures in mind. I recently tried to use TabStripItems and found that they did not behave well in a ScrollViewer such that the selected event was fired when the items were scrolled. It would be useful to be able to select or avoid controls depending on the target platform.

**Reply by timunie** [reply]
I don't think such a list exists. One indicator is the time it is in Avalonia, but not a good indicator. We have @emmauss working on mobile support who already fixed a lot of mobile issues (you see mobile is seriously on our list). For now the main focus is on Text input as this is one of the key features in our understanding. 

Nonetheless, if you find some control can be imporved for mobile, file a feature request telling your needs. For the Tabstip I belive a Style would already solve a lot. Now that we have a platform selector, that should not be a huge undertaking.

**Comment by buzzware** [comment]
The most full featured mobile friendly Avalonia UI framework seems to be using https://github.com/AvaloniaCommunity/Material.Avalonia. See the demo for a hamburger menu, dialogs, snackbar/toasts etc @SKProCH

**Reply by timahrentlov** [reply]
It would be nice to have some sort of official mobile roadmap.

**Reply by MikeCodesDotNET** [reply]
@timahrentlov, we keep our roadmap as an exclusive benefit for our paying customers. 

If there's something specific you'd like to see, feel free to share your thoughts, as we value community feedback and use it to guide our plans for the future.

**Reply by timahrentlov** [reply]
> @timahrentlov, we keep our roadmap as an exclusive benefit for our paying customers. 
> 

Ah. I'm afraid I'm only a small donor and have no need for XPF or such. But I'm glad to learn that roadmaps exist. Perhaps even for mobile.

**Reply by timunie** [reply]
Xpf is not paid support and much cheaper for individual devs. Just in case. Donations are also really welcome, not that you got me wrong.

**Comment by CodingOctocat** [comment]
AvaloniaUI vs MAUI in 2024? I'm currently developing a simple program using MAUI (something like Todo list + watermarked camera), MAUI has a lot of small bugs and some more serious issues, plus it's missing some controls, I don't know what's going on with AvaloniaUI at the moment, and I think it's always worth opening this discussion.

**Reply by wsficke** [reply]
AvaloniaUI if you need cross-platform, WinUI 3 + CommunityToolkit + DynamicData (no MAUI) if you are targeting windows desktop (WPF is also old but good).  MAUI is in a transitive state with successor code of Xamarin but per my evaluation (about a year ago) had quality control issues.

If you are coding for work and you go with Avalonia, get the paid support; Avalonia has its own quirks.  Avalonia is a good tool but sometimes there is more impedance to find the correct patterns than one would like, and the documentation is needing improvement,

**Reply by buzzware** [reply]
I'm using Avalonia for mobile and its working ok from a purely functional perspective for an internal business app. I'm using the Material package (https://github.com/AvaloniaCommunity/Material.Avalonia) which you pretty much need to implement mobile UI. I have my own router (https://github.com/buzzware/PelicanDotNet) and data layer library (https://github.com/buzzware/cascade).
Avalonia have intentions for better mobile UI support, but WPF for desktop apps is their # 1 focus.
I'm also a Flutter developer, and no dotnet mobile framework even comes close to it or React Native for app store apps.

**Reply by timahrentlov** [reply]
My 2 cents: Jep. You'll need a plan b for mobile as things stand right now. I've started learning SwiftUI. Perhaps [skip](https://skip.tools) will come through for Android. Perhaps not. I don't care. As long as I can target the Apple-verse.

**Reply by davidhollitervis** [reply]
We had to make a change and decision at the beginning of the year and chose to go with Avalonia, as we previously used WPF for our desktop apps. I have to say, coming from Xamarin and also after the disappointment with MAUI, I am more than happy with that decision

**Reply by timahrentlov** [reply]
Avalonia for desktop rocks!

**Reply by azegallo** [reply]
I ported a pretty complex project with a large user base from WPF to Avalonia. 
I've since added Android and iOS projects (that only took a couple weeks 😁). 
I'm experimenting with a WASM version (at least a lite version until dotnet fixes WASM threading).

I avoided using any 3rd party libs that don't show a lot of staying power / quick Avalonia updates.

The biggest problem is definitely the docs. But overall I haven't come across anything I couldn't get done with some long searches or help in the Telegram. 

For desktop: I am full send. It's so good. 

For mobile: I wish there were more built in controls, and I've come across some minor bugs, but again I've been able to do everything I've needed so far with a little elbow grease. If I was building a mobile only app I would probably use another framework at this point but if desktop was also needed I'm happy to use Avalonia. This will probably change soon as Avalonia adds more mobile effort (or I keep building out my own mobile tools). But if I needed mobile+desktop.. Avalonia, no brainer. 

Browser: I'll let you know 🤔
